Add a DeskHog blogpost (#12189)
PostHog/posthog.com: contents/blog/welcome-to-deskhog.mdx

---
title: "DeskHog is here"
date: 2025-07-21
author:
  - danilo-campos
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/deskhog_2888c77766.jpg
featuredImageType: full

> featuredVideo: https://youtu.be/lqeD2vvAA4w

Reviewer: @joethreepwood @daniloc we need to use the no cookie version of youtube URL here

Reviewer 2: ```suggestion
featuredVideo: https://www.youtube-nocookie.com/embed/lqeD2vvAA4w
```

---
title: "Welcome to Deskhog"
date: 2025-07-21
author:
  - danilo-campos
featuredImage: >-

>   https://res.cloudinary.com/dmukukwp6/image/upload/robot_6d2cab1b66.png

Author: oh definitely, just didn't get to it yet

featuredImageType: full
tags:
    - Hardware
---

> 

Reviewer 2: We should throw the video in here too

Author: indeed!

DeskHog is here

I want you to fall in love with hardware.

I love this job because the incentives are awesome: PostHog wins when you build things. The more successful you are as a technical creator, the more we get to grow.

Assuming you still like us when you pick your next project.

Because this is true, I’ve been able to convince whole teams and all our executives to support me in doing something ridiculous: shipping you some hardware.

That hardware is DeskHog, and I can’t wait for you to play with our new developer toy.

## Software, ironically, is the hard bit

> DeskHog packages an off-the-shelf, open source board: [Adafruit’s ESP32-S3 Reverse TFT Feather](https://www.adafruit.com/product/5691). Like all of Adafruit’s line, this board is great: it’s got a screen built in, it’s got some buttons, and built-in battery management. 

Reviewer 2: ```suggestion
DeskHog packages an off-the-shelf, open source board: [Adafruit’s ESP32-S3 Reverse TFT Feather](https://www.adafruit.com/product/5691). Like all of Adafruit’s line, this board is great: it’s got a screen built in, it’s got some buttons, and built-in battery management. The issue for most builders though is that it doesn't have anything else: no apps, no firmware. So, we built one!
```

I think we can lay this out more clearly for the norms

> But man, putting it to use is hard. The internet is packed with code that will demonstrate one feature or another of an ESP32. Actually getting these features working *together* is a whole other problem.
> 
> Compounding this issue, the ESP32-S3 is much more resource-constrained than most developers are used to. You have to plan carefully, or risk crashing the board altogether.
> 
> DeskHog solves this with open source firmware. Our code provides:

Reviewer 2: ```suggestion
Without the DeskHog firmware, putting the board to use is very difficult. The internet is packed with code that will demonstrate one feature or another, but actually getting these features working *together*? That's a whole other story.

Compounding this issue, the ESP32-S3 is much more resource-constrained than most developers are used to. You have to plan carefully, or risk crashing the board altogether.

The firmware we've created is rock-solid, open-source, and provides a bunch of included features:
```

Just making this work with the suggestion above

Author: I'm not going to say "rock solid" because that seems like tempting fate but yes

- Built-in wifi setup
- A web portal for configuration
- A custom UI built on paginated cards, with button handling
- Graphics via the *vast* LVGL project
- Networking examples
- Stability despite all these tasks running simultaneously

## Your bug is your bug

That last detail is essential. There’s so much going on in microcontroller land. It would be easy to drown as a newbie.

That was me a couple years back.

So we’ve done the homework. An ESP32-S3 has two cores: tasks are divided between them, with networking on one, and the user-facing application running on the other. It’s easy to trip yourself on the classic horrors of concurrency this way and crash the board. DeskHog’s solution: an event queue that can safely pass messages between tasks without mangling memory.

> I’ve got a DeskHog sitting on my desk running for days on end without drama.

Reviewer 2: ```suggestion
I’ve had a DeskHog sitting on my desk running for days on end without drama.
```

Which means if you write some new code for DeskHog, you know any crash that shows up is a result of the new code, not some obscure detail lurking in a system you barely understand.

> ## Built for developers – of all experience levels

Reviewer 2: ```suggestion
## Optimized for vibe-coding
```

Always worth revisiting the value prop clearly

> You know what sticks in my craw about example microcontroller code?
> 
> At its worst, it’s just thousands of lines of code in one file. I cannot work like that. I need tidy encapsulation, separation of concerns – and because it *is C++* – an object-oriented approach to building new features.
> 

Reviewer 2: ```suggestion
```

I think we can cut these for better flow

> If you’re like me, you’re going to feel right at home in the DeskHog project. You’ll find folders to organize network code, UI code, network clients and parsers… it’s all laid out the way you’d expect from building a modern application.

Reviewer 2: ```suggestion
If you’re an experienced dev, you’re going to feel right at home in the DeskHog project. You’ll find folders to organize network code, UI code, network clients and parsers… it’s all laid out the way you’d expect from building a modern application. 

Most example microcontroller code is just thousands of line of code in one file and I hate that, so we've made sure DeskHog is **organized**.

```

> But maybe you’re not like me. Maybe you don’t care about any of that, and you just think hardware is neat. We’ve got you covered there, too.

Reviewer 2: ```suggestion
And if you're not an experienced dev? If you don't care about tiny encapsulation and an object-orientated approach? We’ve got you covered there, too.
```

Reviewer 2: I think the key change I'm trying to make here is flipping the script to focus on the **user** and how the features benefit them, rather than leading with what bothers you about the straw-man example. The key message for this section should be: DeskHog works at both skill levels and that's a feature, not that you've got an issue with example code.

> The whole project is built for agent-based LLM coding. The individual class interfaces are heavily commented, and the project itself has loads of detail for you to stuff into an agent’s context. My homie Joe [has a great tutorial]() to get you moving on this.

Reviewer 2: ```suggestion
The whole project is built with agent-based LLMs and vibe-coding in mind. Everything is heavily commented, with robust README files that you can stuff into an agent's context and [Joe](/community/profiles/29070) has [a great tutorial for getting started with DeskHog and Claude Code](/tutorials/deskhog-claude-tutorial).

We even tested DeskHog's vibe-coding capabilities by putting it in the hands of five non-technical team members at our most recent hackathon. Most of them had never used tools like Claude and Cursor before, but within 48 hours they were able to build a library of simple games that are now included with DeskHog as example projects.
```

I think dumbing down the language and upping the ambition here will help it speak to vibecoders

Author: The stock of "vibe coding" has taken a serious dump of late but I'm with you on these. I'll say it without saying it

> I’m so excited to say it works! We tried it out: at PostHog’s last offsite, a whole hackathon team formed to experiment with DeskHog. Folks built their first C++ features with the help of agent coding to huge success.

Reviewer 2: ```suggestion
```
